File: Easy/88.合并两个有序数组.cpp
/*
 * @lc app=leetcode.cn id=88 lang=cpp
 *
 * [88] 合并两个有序数组
 */

// @lc code=start
class Solution
{
public:
    void merge(vector<int> &nums1, int m, vector<int> &nums2, int n)
    {
        int i = m - 1, j = n - 1, k = m + n - 1;
        while (i >= 0 && j >= 0)
        {
            //从后往前按照降序的方法插入元素，这样从前往后就是升序了
            if (nums1[i] < nums2[j])
                nums1[k--] = nums2[j--];
            else
                nums1[k--] = nums1[i--];
        }
        //如果nums2还有升序元素的话，就把剩下的元素全部赋上去
        while (j >= 0)
            nums1[k--] = nums2[j--];
    }
};
// @lc code=end

File: Easy/88.合并两个有序数组.go
/*
 * @lc app=leetcode.cn id=88 lang=golang
 *
 * [88] 合并两个有序数组
 */

// @lc code=start
// 从后往前从大到小填充，可以防止数据丢失
func merge(nums1 []int, m int, nums2 []int, n int) {
	if n == 0 {
		return
	}
	if m == 0 {
		copy(nums1, nums2)
		return
	}
	index := m + n - 1
	for p1, p2 := m-1, n-1; p1 >= 0 || p2 >= 0; index-- {
		if p1 == -1 {
			nums1[index] = nums2[p2]
			p2--
		} else if p2 == -1 {
			nums1[index] = nums1[p1]
			p1--
		} else if nums1[p1] >= nums2[p2] {
			nums1[index] = nums1[p1]
			p1--
		} else {
			nums1[index] = nums2[p2]
			p2--
		}
	}
	return
}

// @lc code=end

